packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,278 @@
+
+    /**
+     * Optional health, percentage number.
+     */
+    readonly health?: number
+
+}
+
+/**
+ * The connection-status states.
+ */
+export enum ConnectionState {
+    ONLINE,
+    OFFLINE
+}
+
+@injectable()
+export class ConnectionStatusOptions {
+
+    static DEFAULT: ConnectionStatusOptions = {
+        requestTimeout: 1000,
+        retry: 5,
+        retryInterval: 2000,
+    };
+
+    /**
+     * Timeout for the HTTP GET request in milliseconds.
+     */
+    readonly requestTimeout: number;
+
+    /**
+     * Number of accepted timeouts.
+     */
+    readonly retry: number;
+
+    /**
+     * Retry interval in milliseconds.
+     */
+    readonly retryInterval: number;
+
+}
+
+@injectable()
+export class FrontendConnectionStatusService implements ConnectionStatusService, FrontendApplicationContribution {
+
+    protected readonly statusChangeEmitter: Emitter<ConnectionStatusChangeEvent>;
+    private readonly endpointUrl: string;
+
+    private connectionState: ConnectionStateMachine;
+    private timer: number | undefined;
+
+    constructor(
+        @inject(ConnectionStatusOptions) @optional() protected readonly options: ConnectionStatusOptions = ConnectionStatusOptions.DEFAULT,
+    ) {
+        this.statusChangeEmitter = new Emitter<ConnectionStatusChangeEvent>();
+        this.endpointUrl = new Endpoint().getRestUrl().toString();
+        this.connectionState = new ConnectionStateMachine({ threshold: this.options.retry });
+    }
+
+    onStart() {
+        this.timer = window.setInterval(() => {
+            const handle = (success: boolean) => {
+                this.connectionState = this.connectionState.next(success);
+                this.statusChangeEmitter.fire(this.connectionState);
+            };
+            const xhr = new XMLHttpRequest();
+            xhr.timeout = this.options.requestTimeout;
+            xhr.onreadystatechange = event => {
+                const { readyState, status } = xhr;
+                if (readyState === XMLHttpRequest.DONE) {
+                    handle(status === 200);
+                }
+            };
+            xhr.onerror = () => handle(false);
+            xhr.ontimeout = () => handle(false);
+            xhr.open('GET', `${this.endpointUrl}/alive`);
+            try {
+                xhr.send();
+            } catch {
+                handle(false);
+            }
+        }, this.options.retryInterval);
+        this.statusChangeEmitter.fire(this.connectionState);
+    }
+
+    onStop() {
+        if (this.timer !== undefined) {
+            window.clearInterval(this.timer);
+            this.timer = undefined;
+        }
+    }
+
+    get onStatusChange() {
+        return this.statusChangeEmitter.event;
+    }
+
+}
+
+/**
+ * Just in case we need to support a bit more sophisticated state transitions than having connection or not.
+ * For instance, `pending`, `reconnecting`.
+ */
+class ConnectionStateMachine implements ConnectionStatusChangeEvent {
```

Comment:
minor: I think it should be re-exported if someone wants to modify defaults by rebinding `FrontendConnectionStatusService `